### PR TITLE
Adds storage related events

### PR DESCRIPTION
**Pull request by headzoo**
Emits events using the internal dispatcher when storage values are set/get.

* `StorageEvents.EVENT_APP_CHANGED` - Emitted after an app value is successfully sent to the API.
* `StorageEvents.EVENT_ENTITY_CHANGED` - Emitted after an entity value is successfully sent to the API.
* `StorageEvents.EVENT_APP_FETCHED` - Emitted after an app value is successfully fetched from the API.
* `StorageEvents.EVENT_ENTITY_FETCHED` - Emitted after an entity value is successfully fetched from the API.

In all instances the event listeners receive the entityId, name, and value. For batch gets/sets the listeners receive the entityId and batch name/value pairs object.

```js
app.on(StorageEvents.EVENT_APP_CHANGED, (entityId, name, value) => {
  if (typeof name === 'object') {
    console.log('Batch update:', name);
  } else {
    console.log('Single update:', name, value);
  }
});
```

These changes will help bridge sdk-core storage values with sdk-react component state, by making it easier to update component/global state when storage values are changed.

### CODE REVIEW

**Comment by helstern** [comment]
@headzoo can you provide a concrete example of the scenario you are describing ?

**Comment by headzoo** [comment]
The goes back to what I mentioned last week about needing state management. This PR is the first step towards introducing a store and state management to sdk-react. The changes I made allow me to bind sdk-core storage to a React store (e.g. Redux). So that updating app settings automatically updates app _state_, which leads to automatic re-rendering of the components. When I change the app settings from the settings "page", all other components should be instantly notified of the changes so they can re-render.

The changes in this PR are part of a bigger plan to overhaul sdk-react a bit. My general impression from using sdk-react to build the Github app, and from going through the Trello app code, is that things aren't really being done the "React way." Currently sdk-react is a _very_ thin wrapper around sdk-core. It doesn't do much to abstract away sdk-core or the "classic javascript" way of doing things. sdk-react is basically just sdk-core with themes. It's missing many of the features that makes working with React such a pleasure.

There's more to come after this PR. For instance all the UI stuff should be stored in app state. I shouldn't have to call `dpapp.ui.isLoading()`. Components should automatically re-render when the app transitions from a "loading" state to a "ready" state. I shouldn't have to add code like `dpapp.on('app.refresh', () => .. )` to my app. This stuff should be automatic. Update the state and the components automatically re-render. There are a lot of little things that should be changed, and a lot of PR are on the way.

**Comment by helstern** [comment]
@headzoo I think i can see what you're trying to do, but it's very hard to follow your direction without any knowledge of the big plan.

In my oppinion there should be a clear separation between ui state (managed by react-sdk) and app state (managed by user), and the decision to update ui state after an app state update should belong to the user.

On adding code like `dpapp.on('app.refresh', () => .. )` i think you are wrong though, because apps might choose themselves how to handle when the user hits the refresh button, and i think this should not be constrained or hidden away.

**Comment by chroder** [comment]
Just to chip in here -- the flow of state should be more seamless as you say, but a few points:

* The default SDK _should not_ use redux or any other state management. Redux is often _too much_ for simple things. A lot of people will absolutely write apps using "classic javascript" methodology.
* Remember also that the react SDK is not the only one we want to support. We will ship "simpler" SDKs (e.g. jquery) whose target audience are web dev hackers.
* To that ends, events like app.refresh etc should all still remain. I don't care if the flow of state means values gets disseminated to components in a cleaner way when using react, but "low level" event needs to stay.

Finally, we need to keep an eye on time/goals. I think we need to aim to get GitHub shipped by the end of the week. If not everything can be super clean and sexy, then so be it; we just need to have forward movement or else we'll get stuck in a never ending cycle of refactoring or rewriting. We can continue to make improvements as we go; the point is we really just need to ship something.

**Comment by chroder** [comment]
Anyway, PR looks good IMO -- just wanted to take that ^ opportunity to make sure we all remember the target audiences.

**Comment by headzoo** [comment]
@chroder 

> To that ends, events like app.refresh etc should all still remain. 

I'm not suggesting anything else. sdk-core is fine the way it is, and I understand you're supporting many SDKs, which is why sdk-core exists separately from sdk-react. The changes I'm suggesting would be added to the React boilerplate, with maybe a few minor changes to sdk-react. I'm proposing adding a _bit more_ abstraction to the boilerplate so that React developers can apply React methodologies when writing React apps.

> We will ship "simpler" SDKs (e.g. jquery) whose target audience are web dev hackers.

Then there's no point in hindering sdk-react in order to make it easier to use by devs who probably aren't going to be using it.

The point of having multiple SDKs is making it easier to write DeskPRO apps by letting devs use their existing experience -- those devs most familiar with jQuery will use sdk-jquery, and those most familiar with React will use sdk-react. But as it stands both SDKs would be nothing more than thin wrappers around sdk-core. Which means both jQuery and React devs have to learn sdk-core in order to write apps instead of using their existing knowledge. Which means jQuery devs wouldn't much enjoy using sdk-jquery, and React devs wouldn't much enjoy using sdk-react.

I mean, if you're not following the best practices for jQuery and React in their respective SDKs, and they're just basically going to be the same, then there's almost no point in having multiple SDKs. Why would I, a React developer, even want to use sdk-react / boilerplate if I don't get to leverage the power of React?

> The default SDK should not use redux or any other state management. Redux is often too much for simple things.

I don't entirely agree, but I think you're missing the point that state management would be used internally by sdk/boilerplate, and the average developer would never be exposed to it. All they need to know is that props get updated when changes are made. How those changes happen under the hood is an implementation detail they don't need to worry about.

In the same way I'm "hiding" the use of React Router. Most devs will never know or care how their app switches from one page to another. It's all handled internally, but by using React Router internally, the advanced developer writing an advanced app can open the hood and do more advanced stuff.

>  A lot of people will absolutely write apps using "classic javascript" methodology.

Those developers should probably use sdk-core instead.

> Finally, we need to keep an eye on time/goals.

I agree.

@helstern 

> In my oppinion there should be a clear separation between ui state

I would agree if we weren't using React, but mixing those concerns is very much the "React way" of doing things. React has its own opinions on how UI should be handled, and they differ from yours. React methodologies preach the use of a single "store" -- a single source of truth -- which stores all app values: ui state, settings, form values, etc. Which is what gives React its power. When I change my Github access key on the settings "page", the component displaying my Github avatar automatically re-renders, and the component displaying a list of repos automatically re-renders, and so on. You're kind of taking all the power out of React when it's not used how it was designed to be used.

> because apps might choose themselves how to handle when the user hits the refresh button

I completely agree, but I'm saying there are _React ways_ of doing that kind of stuff. React developers don't usually listen for UI changes with event handlers. That's the jQuery way of doing things. In React the UI is described by the state tree, and the UI is updated by modifying the state. Developers are free to ignore state changes by using the `componentWillUpdate()` method, but binding event listeners to the UI is kind of "code smell" in the React world.

**Comment by chroder** [comment]
Like I said, this PR looks good, I don't have anything against it :-) You seem to think I'm disagreeing or something.

My point is to limit the how far down the rabbit hole this can go. E.g. I don't want redux knowledge to be a prerequisite to build an app in React. You can disagree all you want, but the fact remains that a significant portion of React developers are not using redux; indeed a significant portion of react developers use React as-is.

So hide implementation details etc -- all good. Just avoid adding entire methodologies or frameworks because that just increases learning curve. We can always release a more opinionated "Redux" boilerplate later if we want to.

Obviously this PR is not doing that. I'm just pointing it out so we're all on the same page.

Edit: To clarify again -- it doesn't matter what is going on "under the hood"; obviously it doesn't matter what's going on under the hood if it's not in a users face. I just don't want to assume redux/flux is what everyone uses and what everyone knows -- because that is simply not the case. Remember the target audience is junior/mid level web devs who want to whip together an app in 30 minutes by performing some API call. It's not a 100k LOC redux app and never will be.

**Comment by helstern** [comment]
@headzoo, thank you for explaining React to me :)), but I don't think my views conflict with / exclude React.
I just think we should not assume too much on behalf of sdk-user. Should be up to them to hook up a store etc, we should facilitate this, but not force the choice

Regarding this PR, there is nothing technically wrong with it, but if we're not going to add the store changes etc, i think we should not merge it yet. i leave it to @chroder to break the tie

**Comment by headzoo** [comment]
@helstern @chroder 

>  but it's very hard to follow your direction without any knowledge of the big plan.

You guys seem to think I'm trying to make the sdk more complicated, when I'm actually trying to make it much, much easier to use. The level of complexity in the Trello app kind of blows my mind, so in the interest of "Clean up framework" I started thinking about ways to simplify things.

This is kind of how I envision the sdk-react and boilerplate using very rough pseudocode, but it shows how the main component of the app doesn't need to mess with sdk-core (e.g. dpapp), and Redux (or whatever) is hidden away in sdk-react. The end user of the sdk/boilerplate never sees it. They only need to worry about prop values. Which, despite using Redux (or whatever) actually makes the sdk/boilerplate **easier** to use. It reduces the learning curve because noobs only have to deal with prop changes instead of sdk-core.

```jsx
// -- index.js
//
// This code would be in the boilerplate.

import ReactDOM from 'react-dom';
import { DeskproAppContainer } from '@deskproapps/deskproapps-sdk-react';
import App from './App';

export function runApp(app, manifest) {
  ReactDOM.render(
    <DeskproAppContainer app={app} manifest={manifest} name={'Your app name'}>
      {/*
        Pass the app as a child of the container so that developers can initialize it
        how they see fit. This follows a standard pattern used by React devs.
      */}
      <App />
    </DeskproAppContainer>,
    document.getElementById('deskpro-app')
  );
}
```

```jsx
// -- DeskproAppContainer.js
//
// This code would be hidden away in the react sdk. This is not
// code the end user writes. This is the "abstraction" that reduces the amount of
// boilerplate the developer needs to write.

import React from 'react';
import {connect} from 'react-redux';
import { StorageEvents } from '@deskproapps/deskproapps-sdk-core';
import { storageChanged } from 'actions/storageActions';

class DeskproAppContainer extends React.Component {
  componentDidUpdate() {
    const { dpapp, dispatch, manifest } = this.props;
    
    // Dispatch the fetched storage values to the Redux store.
    dpapp.on(StorageEvents.EVENT_APP_FETCHED, (entityId, name, value) => {
      dispatch(storageChanged(name, value));
    });
    dpapp.on(StorageEvents.EVENT_APP_CHANGED, (entityId, name, value) => {
      dispatch(storageChanged(name, value));
    });

    // Automatically fetch any values declared in the manifest. I think the first thing
    // most apps will do is fetch values from storage, which means it should be done
    // automatically by the sdk.
    //
    // There's no need for write .then().catch() here, because the fetched values will
    // be automatically dispatched to the store by the above event listener. When the store
    // receives the changes, it will call render() again with new props, which will get
    // passed down to the "main component".
    Object.keys(manifest.state).forEach((key) => {
      dpapp.storage.getAppStorage(key);
    });
  }
  
  render() {
    // The children here is the "main component". It will receive -- as props -- each storage
    // value declared in the manifest. Instead of the main component calling
    // dpapp.storage.getAppStorage('settings'), it would simply read from this.props.settings.
    const { children, ...props } = this.props;
    return React.cloneElement(children, props);
  }
}

function mapStateToProps(state) {
  // The state would contain (among other things) the storage values
  // declared in the manifest. These state values become props in the
  // container which get passed down to the "main component."
  return Object.assign({}, state);
}
export default connect(mapStateToProps)(DeskproAppContainer);
```

```jsx
// -- App.jsx
//
// This is the app or "main component." Note that it doesn't need to muck
// about with sdk-core at all. It has already described the storage values
// in the manifest, which get passed to it by props. When the storage values
// change, the props change, and components _can_ re-render if needed.
//
// This is _basically_ the only code the end user needs to write, which is a good deal
// more simple than what we see in the existing apps.

import React from 'react';
import PropTypes from 'prop-types';

const App = ({ settings }) => (
  <div>
    <img src={settings.avatar} />
  </div>
);

App.propTypes = {
  settings: PropTypes.object.isRequired
};

export default App;
```

**Comment by chroder** [comment]
That's a definite improvement yeah. There are some things to resolve still though:

a) There's a major issue here around forcing redux. If a developer does not use redux, then this could result in some pretty funky behaviour where an update causes the root container to re-render, and thereby possibly affecting any children that might not get state from redux.

b) There are some missing bits that would need to be implemented, like how to handle clicking settings -- we'd need to expose that some other way, or continue to pass through `dpapp` so the app can attach events "manually".

c) I guess getting app storage keys would be better off using a batch, but that's an implementation detail.

---

Feels like this is `react-redux-sdk`. We really need to keep a small and "dumb" react SDK (again, not everyone wants to use redux!), but I'm 100% okay with adding something more complete and opinionated on top of it.

**Comment by headzoo** [comment]
> There's a major issue here around forcing redux.

I get that, but we're not forcing redux on anyone.

```jsx
import React from 'react';
import PropTypes from 'prop-types';

const App = ({ settings }) => (
  <div>
    <img src={settings.avatar} />
  </div>
);

App.propTypes = {
  settings: PropTypes.object.isRequired
};

export default App;
```

Do you see anything Redux or even store/state related in that code? To the end user the values they define in the manifest magically become props in their app. Simple as that. The devs never touch Redux. They don't even know it's being used.

> There are some missing bits that would need to be implemented

There's more to my idea, but I don't want to go too deep down this discussion if no one else likes the general premise. But I can write up a more complete example later tonight on my own time.

Here's the deal.. I consider myself at least a mid-level Javascript guru at this point, and I can't even follow the code in the Trello app. So I really don't think noobs will be able to follow it either. My goal here is making the SDK easier to use for noobs, and right now it's promise-hell and pretty darn complicated. It's also solving a lot of problems which React already solved, which @chroder eluded to on the Github trello board.

More to the point, most of us dropped jQuery like a hot potato when React came along because we were sick and tired of Javascript/jQuery and binding events and dealing with callbacks and promises. I've never written a promise in a React component in four years, because you don't need them. Working with sdk-react is kind of a drag because it's forcing me back into a methodology that I never want to deal with again. I wouldn't even be using React if I had to continue doing things that way. As a React developer I wouldn't want to use sdk-react because there's not much about it that's "React-ish." Writing the Github app is a pain because doing things which are normally easy with React are not easy with sdk-react, and it's frustrating.

So, to reiterate, my goal here is making sdk-react _usable_ by inexperienced developers. Right now it's way too complicated. Thinking about the poor inexperienced developers who have to use the sdk is exactly what I'm doing.

**Comment by chroder** [comment]
No one is disputing that an improvement should be made; really not sure where you're getting that impression. This discussion is around where it should go, in react-sdk which so far has been "dumb", or a layer on top of it.

From what I'm reading, you're proposing that there's not really any point in having a separate dumb react-sdk because using redux is completely hidden from the user anyway, and doesn't affect them in any way. Is that right?

So the requirements here are thus:

* Redux is invisible (unless you want to use it of course).
* E.g. someone following the [react tutorial](https://facebook.github.io/react/tutorial/tutorial.html) should see no new concepts and won't run into any surprises.
* Using some other library (e.g. MobX or reflux) should be possible with no changes or special considerations on the user.

If all of those things pass (from what you describe, sounds like it should?) then I'm okay with adding redux to the SDK to encapsulate these things and make it all cleaner. If any one of those things doesn't pass, then we should build a lib on top of the dumb sdk instead.

I raise these points because you're in a better position to answer them. So it's a pretty simple clear-cut question, a or b. I don't really care so long as the requirements are met. Make sense?

**Comment by headzoo** [comment]
Yup, you've got it right, although I'm not sure yet if we need something as complex as redux, or if we should piggyback on the event dispatching system already built into sdk-core, and given the time constraints it's not something I plan on messing with this week. Maybe over the weekend I can cobble something together to give you guys a better idea of the grand plan.

As for this PR, I think the changes are worth having even if we don't go down the store/state road. They will help other devs who may want to add a store to their app.

**Comment by chroder** [comment]
OK cool, sounds good. Agreed re this PR, it's mainly been kept open just while we discussed that (albeit somewhat unrelated!).

I'll merge now 👍

**Review by chroder — APPROVED** [review]

**Comment by chroder** [comment]
Or rather @helstern you should merge after approving. Looks OK to me but will leave the approval to you.

**Review by helstern — APPROVED** [review]

**Pull request merged**
